### ISSUE TITLE
spark-submit hangs after successful job completion

### ISSUE THREAD

**Issue by abhivijay89** [bluedenim/log4j-s3-search#76]
Hi @bluedenim 

When executing a job via Spark using the s3-log4j-appender implementation, spark-submit command seems to hang after successful job execution. The logs are getting pushed to s3 based on 'stagingBufferSize' or 'stagingBufferAge' at the intermediate stages of the job execution but hangs at the end after SparkContext gets stopped successfully. I need to manually invoke ctrl-c to terminate the spark-submit process & then it goes ahead and publishes the staging log once the shutdown hook gets triggered. Took a thread dump to troubleshoot and I see this:

```
"LoggingEventCache-publish-thread" #229 prio=5 os_prio=0 tid=0x00007f4315601800 nid=0x4df waiting on condition [0x00007f430c8fa000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000002c014f9c0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Can you help to check what maybe causing this & provide insights on how this could be fixed? Has it got to do anything with the changes we did to publish in the current thread itself when shutdown hook gets invoked...

**Comment by abhivijay89** [comment]
@bluedenim Even after adding `LoggingEventCache.shutDown()` in my program after `spark.stop`, Spark hangs and upon checking the thread dump, I still see that the publish thread is waiting.
If I add `System.exit(0)` as a workaround (not preferable) , then it seems to works fine.
Am I going wrong somewhere/missing something?

**Comment by bluedenim** [comment]
That's odd. I was able to reproduce the problem with a simple program and then verified that the fix worked. Here's the program:
```
public class Main {
    private static Logger logger =
        LogManager.getLogger(Main.class);

    public static void main(String[] args) throws InterruptedException {
        logger.info("Hello from Main!");
        Long started = System.currentTimeMillis();
        Long now = System.currentTimeMillis();

        // Loop for 15 secs
        while (now - started < TimeUnit.SECONDS.toMillis(15)) {
            logger.info("Another round through the loop!");
            logger.warn("This is a warning!");
            logger.error("And this is an error!!!");

            Thread.sleep(TimeUnit.SECONDS.toMillis(2));
            now = System.currentTimeMillis();
        }

        LoggingEventCache.shutDown();
    }
}
```
and its output:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
23:20:51,085 [main] INFO  com.van.example.Main - Hello from Main!
23:20:51,091 [main] INFO  com.van.example.Main - Another round through the loop!
23:20:51,092 [main] WARN  com.van.example.Main - This is a warning!
23:20:51,092 [main] ERROR com.van.example.Main - And this is an error!!!
23:20:53,097 [main] INFO  com.van.example.Main - Another round through the loop!
23:20:53,098 [main] WARN  com.van.example.Main - This is a warning!
23:20:53,098 [main] ERROR com.van.example.Main - And this is an error!!!
23:20:55,107 [main] INFO  com.van.example.Main - Another round through the loop!
23:20:55,107 [main] WARN  com.van.example.Main - This is a warning!
23:20:55,107 [main] ERROR com.van.example.Main - And this is an error!!!
23:20:57,115 [main] INFO  com.van.example.Main - Another round through the loop!
23:20:57,116 [main] WARN  com.van.example.Main - This is a warning!
23:20:57,116 [main] ERROR com.van.example.Main - And this is an error!!!
23:20:59,121 [main] INFO  com.van.example.Main - Another round through the loop!
23:20:59,121 [main] WARN  com.van.example.Main - This is a warning!
23:20:59,122 [main] ERROR com.van.example.Main - And this is an error!!!
23:21:01,137 [main] INFO  com.van.example.Main - Another round through the loop!
23:21:01,137 [main] WARN  com.van.example.Main - This is a warning!
23:21:01,137 [main] ERROR com.van.example.Main - And this is an error!!!
23:21:03,140 [main] INFO  com.van.example.Main - Another round through the loop!
23:21:03,142 [main] WARN  com.van.example.Main - This is a warning!
23:21:03,142 [main] ERROR com.van.example.Main - And this is an error!!!
23:21:05,147 [main] INFO  com.van.example.Main - Another round through the loop!
23:21:05,148 [main] WARN  com.van.example.Main - This is a warning!
23:21:05,148 [main] ERROR com.van.example.Main - And this is an error!!!
LoggingEventCache: shutting down
LoggingEventCache: Executor service terminated within timeout: true

Process finished with exit code 0
```

I'm wondering if in your run you get this when you call `LogginEventCache.shutDown()`:
```
LoggingEventCache: shutting down
LoggingEventCache: Executor service terminated within timeout: true
```

Did yours return false?

**Comment by abhivijay89** [comment]
I don't see these statements printed out. Probably we need to enable verbose mode and then I need to set it to true in log4j?
```
LoggingEventCache: shutting down
LoggingEventCache: Executor service terminated within timeout: XXXX
```

**Comment by abhivijay89** [comment]
@bluedenim I was able to debug and fix the issue on my end. Working like a charm now. Thanks so much for your help!
```
[2020-08-28 12:23:45.884] [INFO ] [spark-listener-group-shared] [XXXX] [com.XXXX.spark.util.AppLifeCycleSparkListener.onApplicationEnd@22] - ApplicationEnd appId=application_XXXX
LoggingEventCache: shutting down
LoggingEventCache: Executor service terminated within timeout: true
Publishing staging log on shutdown...
Collecting content into /tmp/toBePublished7929854388079246499.tmp before uploading.
```

Would you be able to merge this PR and roll out a new release soon?

**Comment by abhivijay89** [comment]
Can you help to review & sneak in this small change as well: https://github.com/bluedenim/log4j-s3-search-samples/pull/2 ?

**Comment by bluedenim** [comment]
Good to hear. Yeah. I'll be able to do a release over the weekend.

**Comment by bluedenim** [comment]
Both https://github.com/bluedenim/log4j-s3-search/pull/78 and https://github.com/bluedenim/log4j-s3-search/pull/77 (this issue) will be in the next release.

**Comment by bluedenim** [comment]
Released 2.8.0. It has this fix.
